fix(vmrestore): volumerestorepolicy inplace flaky tests

SkalaNetworks opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does

This PR fixes flaky tests at https://github.com/kubevirt/kubevirt/issues/14982

#### Before this PR:

Flaky tests on VM restores with VolumeRestorePolicy InPlace

#### After this PR:

Hopefully no flaky test

### References
<!-- optional,
  Use `Fixes #<issue number>(, Fixes #<issue_number>, ...)` format, to close the issue(s) when PR gets merged.
  Use `Partially addresses #<issue number>` to link an issue without closing it when the PR merges.
  
- Fixes https://github.com/kubevirt/kubevirt/issues/14982

-->
<!-- optional,
  VEP tracking issue if this PR is implementing one.
  For additional info about VEP tracking issue, see https://github.com/kubevirt/enhancements#process
  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issue/<vep_tracking_issue_number>
-->

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

My analysis of the flaky test is a possible condition where the PVC deletion may error out, preventing the status of the VMRestore to be updated with the datavolume. But on the next resync cycle, the PVC got deleted because it was held by a finalizer (VM still up, Pod still up, anything that may slow down PVC deletion). We enter the bit of code where DataVolume is expected eventhough it was never set.

To fix this, we exit the reconciliation when setting the property, and we re-enter the reconciliation function on the next loop with a guarantee that the property is set.

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- x ] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

dhiller commented:
/test pull-kubevirt-generate

SkalaNetworks commented:
@dhiller Can I do anything to fix that? https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/14985/pull-kubevirt-check-dequarantine-test/1937416815042891776 this job seems to be running it quite a lot

dhiller commented:
> @dhiller Can I do anything to fix that? [prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/14985/pull-kubevirt-check-dequarantine-test/1937416815042891776](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/14985/pull-kubevirt-check-dequarantine-test/1937416815042891776) this job seems to be running it quite a lot

No, it doesn't seem to run it, see here: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/14985/pull-kubevirt-check-dequarantine-test/1937416815042891776#1:build-log.txt%3A2075

The number 0 / 1395 basically means that no tests have been executed - probably the lane is missing configuration to pick this test up. @akalenyu do you have an idea why that is?

mhenriks commented:
/unhold

ShellyKa13 commented on pkg/storage/snapshot/restore.go:
```diff
@@ -495,7 +496,11 @@ func (ctrl *VMRestoreController) reconcileVolumeRestores(vmRestore *snapshotv1.V
 				log.Log.Object(vmRestore).Infof("marking datavolume %s/%s as prepopulated before deleting its PVC", vmRestore.Namespace, ownerDV)
 
-				vmRestore.Status.Restores[i].DataVolumeName = &ownerDV
+				if vmRestore.Status.Restores[i].DataVolumeName == nil {
```
I think its better here to do the update and mark waitingDVNameUpdate = true or something like that and then return that value at the end same as done for `waitingPVC` for example, so all the relevant restores will be updated in one time.
Keeping it like this will cause a new reconcile for each dv

ShellyKa13 commented on pkg/storage/snapshot/restore_test.go:
```diff
@@ -877,6 +877,7 @@ var _ = Describe("Restore controller", func() {
 						VolumeName:                diskName,
 						PersistentVolumeClaimName: "alpine-dv", // Name is identical to the original PVC
 						VolumeSnapshotName:        "vmsnapshot-snapshot-uid-volume-disk1",
+						DataVolumeName:            pointer.P(diskName), // This will be set at some point by the controller
```
Do we have a test that makes sure this is updated at some point, i.e restore if without this before the processVMRestoreWorkItem and is added after? need to also show that there is no update of the dv with the annotation before the restore status is updated, and also that the pvc is not deleted yet.

SkalaNetworks replied:
now we do :)

ShellyKa13 commented on tests/storage/restore.go:
```diff
@@ -1848,19 +1848,22 @@ var _ = Describe(SIG("VirtualMachineRestore Tests", func() {
-					return restoredDV.Annotations[cdiv1.AnnPrePopulated]
-				}, 60*time.Second, 1*time.Second).Should(Equal(originalPvcName))
+				// DV has the prePopulated annotation with the name of the PVC
+				restoredDV, err := virtClient.CdiClient().CdiV1beta1().DataVolumes(vm.Namespace).Get(context.Background(), originalPvcName, metav1.GetOptions{})
```
you already do this get in line 1858 and also the
Expect(dv.Annotations["restore.kubevirt.io/name"]).To(Equal(restore.Name))

ShellyKa13 reviewed (commented):
Looks good overall, some comments and improvment.
Did you get to run this test several times locally to see it fixes the flake?
Thanks for working on this!

SkalaNetworks commented:
> Looks good overall, some comments and improvment. Did you get to run this test several times locally to see it fixes the flake? Thanks for working on this!

@ShellyKa13 
I ran them about 20 times locally and never encountered an issue, but again, I also didn't get those problems before the flakes popped up on Kubevirt's CI/CD. So I wouldn't say it means much.

I should have addressed your comments. I made the light test that only checked if the restore started a bit better. It now checks the DV is correctly marked in the status, no deletion occurs and no update of the DV is done.

mhenriks commented:
/test pull-kubevirt-e2e-k8s-1.32-sig-storage
/test pull-kubevirt-e2e-k8s-1.31-sig-storage
/test pull-kubevirt-e2e-k8s-1.30-sig-storage

mhenriks commented:
/test pull-kubevirt-e2e-k8s-1.32-sig-storage
/test pull-kubevirt-e2e-k8s-1.31-sig-storage
/test pull-kubevirt-e2e-k8s-1.30-sig-storage

mhenriks commented:
/approve

ShellyKa13 commented on tests/storage/restore.go:
I believe this is not rebased on main? since test 
It("should restore with volume restore policy InPlace and DV (not template) as disk", func() { 
is quarantined on main.. so if you dont remove the quarantine label in this PR it will remain quarantined.

ShellyKa13 commented:
/test pull-kubevirt-e2e-k8s-1.32-sig-storage
/test pull-kubevirt-e2e-k8s-1.31-sig-storage
/test pull-kubevirt-e2e-k8s-1.30-sig-storage

ShellyKa13 commented:
There is some known issue with some lanes which seems like it affects the flakes lane.. not sure if restest will help but lets keep trying?
/test pull-kubevirt-check-tests-for-flakes

ShellyKa13 commented:
/unhold
